File: src/main/java/pers/lyks/spring/example/strategy/CalculateContext.java
package pers.lyks.spring.example.strategy;

import pers.lyks.spring.example.holder.SpringContextHolder;

import java.util.Map;

public class CalculateContext {
    private final Map<String, Class<? extends CalculateStrategy>> handlerMap;

    public CalculateContext(Map<String, Class<? extends CalculateStrategy>> handlerMap) {
        this.handlerMap = handlerMap;
    }

    public CalculateStrategy getInstance(String type) {
        Class clazz = handlerMap.get(type);
        if (null == clazz) {
            throw new IllegalArgumentException("not found handler for type: " + type);
        }
        return (CalculateStrategy) SpringContextHolder.getBean(clazz);
    }
}

File: src/main/kotlin/pers/lyks/spring/example/strategy/CalculateContext.kt
package pers.lyks.spring.example.strategy

import pers.lyks.spring.example.holder.SpringContextHolder

class CalculateContext(private val handlerMap: Map<String, Class<out CalculateStrategy<*, *>>>) {

    fun <F : Number, S : Number> getInstance(type: String): CalculateStrategy<F, S> {
        val clazz = handlerMap[type] ?: throw IllegalArgumentException("not found handler for type: $type")
        return SpringContextHolder.getBean(clazz) as CalculateStrategy<F, S>;
    }
}
